scripts/Diffraction/single_crystal/base_sx.py

```diff
@@ -0,0 +1,574 @@
+        try:
+            return self.runs[str(run)][fieldname]
+        except:
+            return None
+
+    @staticmethod
+    def has_sample(ws):
+        return BaseSX.retrieve(ws).sample().getMaterial().numberDensity > 1e-15
+
+    # --- setters ---
+
+    def set_van_ws(self, van_ws):
+        self.van_ws = BaseSX.retrieve(van_ws).name()
+
+    def set_ws(self, run, ws):
+        run_str = str(run)
+        if run_str not in self.runs:
+            self.runs[run_str] = {"ws": BaseSX.retrieve(ws).name()}
+        else:
+            self.runs[run_str]["ws"] = BaseSX.retrieve(ws).name()
+
+    def set_peaks(self, run, peaks, peak_type=PEAK_TYPE.FOUND, integration_type=None):
+        if integration_type is None:
+            fieldname = peak_type.value
+        else:
+            fieldname = "_".join([peak_type.value, integration_type.value])
+        self.runs[str(run)][fieldname] = BaseSX.retrieve(peaks).name()
+
+    def set_md(self, run, ws_md):
+        self.runs[str(run)]["MD"] = BaseSX.retrieve(ws_md).name()
+
+    def set_sample(self, **kwargs):
+        self.sample_dict = kwargs
+
+    def set_goniometer_axes(self, *args):
+        """
+        :param axes: collection of axes - e.g. ([x,y,z,hand], [x,y,z,hand]) - in same order to be passed to SetSample
+        """
+        # convert to strings
+        self.gonio_axes = [",".join([str(elem) for elem in axis]) for axis in args]
+
+    def set_mc_abs_nevents(self, nevents):
+        self.n_mcevents = nevents
+
+    # --- abstract methods ---
+
+    @abstractmethod
+    def process_data(self, runs: Sequence[str], *args):
+        raise NotImplementedError()
+
+    @abstractmethod
+    def load_run(self, runno):
+        raise NotImplementedError()
+
+    @abstractmethod
+    def process_vanadium(self):
+        raise NotImplementedError()
+
+    # --- methods ---
+
+    def delete_run_data(self, run, del_MD=False):
+        run = str(run)
+        fields = ["ws", "MD"] if del_MD else ["ws"]
+        for field in fields:
+            if self.runs[run][field]:
+                mantid.DeleteWorkspace(self.runs[run][field])
+                self.runs[run][field] = None
+
+    def _set_goniometer_on_ws(self, ws, angles):
+        """
+        :param ws: workspace or name
+        :param angles: list of angles or motor names
+        :return:
+        """
+        if self.gonio_axes is not None:
+            axis_dict = {f"Axis{iax}": ",".join([str(angles[iax]), self.gonio_axes[iax]]) for iax in range(len(angles))}
+            mantid.SetGoniometer(Workspace=ws, EnableLogging=False, **axis_dict)
+
+    @staticmethod
+    def convert_ws_to_MD(wsname, md_name=None, frame="Q (lab frame)"):
+        if md_name is None:
+            md_name = wsname + "_MD"
+        xunit = BaseSX.get_xunit(wsname)  # get initial xunit
+        BaseSX._normalise_by_bin_width_in_k(wsname)  # normalise by bin-width in K = 2pi/lambda
+        wsMD = mantid.ConvertToDiffractionMDWorkspace(
```

Comment:
Does this call leave a PreprocessedDetectorsWS behind? Might be worth cleaning it up at the end of the function